src/app/clusters/boolean-state-configuration-server/boolean-state-configuration-server.cpp

```diff
@@ -0,0 +1,443 @@
+    }
+
+    switch (aPath.mAttributeId)
+    {
+    case CurrentSensitivityLevel::Id: {
+        return ReadCurrentSensitivityLevel(aPath, aEncoder);
+    }
+    default: {
+        break;
+    }
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BooleanStateConfigAttrAccess::Write(const ConcreteDataAttributePath & aPath, AttributeValueDecoder & aDecoder)
+{
+    if (aPath.mClusterId != BooleanStateConfiguration::Id)
+    {
+        return CHIP_ERROR_INVALID_PATH_LIST;
+    }
+
+    switch (aPath.mAttributeId)
+    {
+    case CurrentSensitivityLevel::Id: {
+        return WriteCurrentSensitivityLevel(aPath, aDecoder);
+    }
+    default: {
+        break;
+    }
+    }
+
+    return CHIP_NO_ERROR;
+}
+} // namespace
+
+static bool emitAlarmsStateChangedEvent(EndpointId ep)
+{
+    if (!HasFeature(ep, BooleanStateConfiguration::Feature::kAudible) &&
+        !HasFeature(ep, BooleanStateConfiguration::Feature::kVisual))
+    {
+        return false;
+    }
+
+    BooleanStateConfiguration::Events::AlarmsStateChanged::Type event;
+    BitMask<BooleanStateConfiguration::AlarmModeBitmap> active;
+    VerifyOrReturnValue(EMBER_ZCL_STATUS_SUCCESS == AlarmsActive::Get(ep, &active), false);
+    event.alarmsActive = active;
+
+    if (HasFeature(ep, BooleanStateConfiguration::Feature::kAlarmSuppress))
+    {
+        BitMask<BooleanStateConfiguration::AlarmModeBitmap> suppressed;
+        VerifyOrReturnValue(EMBER_ZCL_STATUS_SUCCESS == AlarmsSuppressed::Get(ep, &suppressed), false);
+        event.alarmsSuppressed.SetValue(suppressed);
+    }
+
+    EventNumber eventNumber;
+
+    CHIP_ERROR error = LogEvent(event, ep, eventNumber);
+
+    if (CHIP_NO_ERROR != error)
+    {
+        ChipLogError(Zcl, "Unable to emit AlarmsStateChanged event [ep=%d]", ep);
+        return false;
+    }
+
+    ChipLogProgress(Zcl, "Emit AlarmsStateChanged event [ep=%d]", ep);
+    return true;
+}
+
+static CHIP_ERROR emitSensorFaultEvent(EndpointId ep, BitMask<BooleanStateConfiguration::SensorFaultBitmap> fault)
+{
+    BooleanStateConfiguration::Events::SensorFault::Type event;
+    EventNumber eventNumber;
+
+    event.sensorFault = fault;
+
+    CHIP_ERROR error = LogEvent(event, ep, eventNumber);
+
+    if (CHIP_NO_ERROR != error)
+    {
+        ChipLogError(Zcl, "Unable to emit SensorFault event [ep=%d]", ep);
+        return error;
+    }
+
+    ChipLogProgress(Zcl, "Emit SensorFault event [ep=%d]", ep);
+    return CHIP_NO_ERROR;
+}
+
+static CHIP_ERROR StoreCurrentSensitivityLevel(EndpointId ep, uint8_t level)
+{
+    uint8_t supportedSensLevel;
+    VerifyOrReturnError(EMBER_ZCL_STATUS_SUCCESS == SupportedSensitivityLevels::Get(ep, &supportedSensLevel),
+                        CHIP_IM_GLOBAL_STATUS(Failure));
+    VerifyOrReturnError(supportedSensLevel >= kMinSupportedSensitivityLevels, CHIP_IM_GLOBAL_STATUS(ConstraintError));
+    VerifyOrReturnError(supportedSensLevel <= kMaxSupportedSensitivityLevels, CHIP_IM_GLOBAL_STATUS(ConstraintError));
```

Comment:
ConstraintError is for when the _incoming data_ is violating the spec.  But here the incoming data is fine; it's our config that's broken.  This should probably just be Failure; it's an internal failure in the server code, not actionable by the client.